ppx/ppx_js/lib_internal/ppx_js_internal.ml

```diff
@@ -723,116 +768,116 @@ let literal_object self_id (fields : field_desc list) =
-                | `Error e -> Exp.extension e
-              in
-              mapper.expr mapper { new_expr with pexp_attributes }
-          | _ -> Ast_mapper.default_mapper.expr mapper expr
-        in
-        default_loc := prev_default_loc;
-        new_expr)
-  }
 
-let () =
-  Driver.register
-    ~name:"ppx_js"
-    Migrate_parsetree.Versions.ocaml_408
-    (fun _config _cookies -> mapper)
+              method_call ~loc ~apply_loc:expr.pexp_loc obj (meth_str, meth.pexp_loc) args
+            in
+            self#expression { new_expr with pexp_attributes }
+        (* obj##meth arg1 arg2 .. *)
+        | { pexp_desc = Pexp_apply (([%expr [%e? obj] ## [%e? meth]] as prop), args)
+          ; pexp_loc
+          ; _
+          } ->
+            let meth_str = exp_to_string meth in
+            let obj = self#expression obj in
+            let args = List.map args ~f:(fun (s, e) -> s, self#expression e) in
+            let new_expr =
+              method_call
+                ~loc:prop.pexp_loc
+                ~apply_loc:pexp_loc
+                obj
+                (meth_str, meth.pexp_loc)
+                args
+            in
+            self#expression { new_expr with pexp_attributes }
+        (* obj##meth *)
+        | [%expr [%e? obj] ## [%e? meth]] as expr ->
+            let obj = self#expression obj in
+            let meth_str = exp_to_string meth in
+            let new_expr =
+              method_call
+                ~loc:expr.pexp_loc
+                ~apply_loc:expr.pexp_loc
+                obj
+                (meth_str, meth.pexp_loc)
+                []
+            in
+            self#expression { new_expr with pexp_attributes }
+        (* new%js constr] *)
+        | [%expr [%js [%e? { pexp_desc = Pexp_new constr; _ }]]] ->
+            let new_expr = new_object constr [] in
+            self#expression { new_expr with pexp_attributes }
+        (* new%js constr arg1 arg2 ..)] *)
+        | { pexp_desc =
+              Pexp_apply ([%expr [%js [%e? { pexp_desc = Pexp_new constr; _ }]]], args)
+          ; _
+          } ->
+            let args = List.map args ~f:(fun (s, e) -> s, self#expression e) in
+            let new_expr = new_object constr args in
+            self#expression { new_expr with pexp_attributes }
+        (* object%js ... end *)
+        | [%expr [%js [%e? { pexp_desc = Pexp_object class_struct; _ }]]] ->
+            let fields =
+              preprocess_literal_object self#expression class_struct.pcstr_fields
+            in
+            let new_expr =
+              match fields with
+              | `Fields fields -> literal_object class_struct.pcstr_self fields
+              | `Error e -> Exp.extension e
+            in
+            self#expression { new_expr with pexp_attributes }
+        | _ -> super#expression expr
+      in
+      default_loc := prev_default_loc;
+      new_expr
+  end
+
+let () = Driver.register_transformation "ppx_js" ~impl:transform#structure
+
+let mapper =
```

Comment:
note that the mapper will not behave correctly if one decide to update other fields. I think we should document it.

Reply:
I've documented that in the .mli-file now. Is that what you were referring to? Or would you prefer a comment in the .ml-file?

Reply:
Yes, I would add a comment in `ppx_js.mli` as it's a public interface.

Reply:
Good point. I've moved the comment from `ppx_js_internal.mli` to  `ppx_js.mli` now.